### PR TITLE
Bug fix for issue #5023 :  corner-case bug in single_source_dijkstra

### PR DESCRIPTION
**Pull request by divyanx**
Fixes issue #5023 and similar issue
`single_source_dijkstra` and `single_source_bellmanford` now raise exception when source and target are same but they are not in graph, which previously was returning 0.

### CODE REVIEW

File: networkx/algorithms/shortest_paths/weighted.py
Comment (dschult) [reply networkx/algorithms/shortest_paths/weighted.py]: Yes -- t hat is correct... no need to add this check to multi source dijkstra

**Review by dschult — APPROVED** [review]
This looks ready to me... :}

File: networkx/algorithms/shortest_paths/weighted.py
Comment (rossbar) [inline networkx/algorithms/shortest_paths/weighted.py:228]:
```suggestion
```

**Comment by divyanx** [comment]
> This looks ready to me... :}

Hey @dschult and @MridulS , can you also merge it, if everything is fine. That will be very helpful to me.

**Review by rossbar — COMMENTED** [review]
+1 for all of the checks in the `if source == target` statements. The one sticking point is the check that was added to `multi_source_dijkstra`, which is currently redundant.

I think the easiest thing to do would be to back that one change out here and handle it separately. Everything else is an obvious fix and should go in!

File: networkx/algorithms/shortest_paths/weighted.py
Comment (dschult) [reply networkx/algorithms/shortest_paths/weighted.py]: OKay -- it appears we need to be more clear about the `multi_source` case.  The function `_djikstra_multisource` should **not** have any checks in it -- they should be done in the functions that call `_djikstra_multisource`.
The function `multi_source_djikstra` **should** have the check just after it checks the sources input (as you have done it).  But the function `multi_source_djikstra_path_length` should **also** check its inputs.

So, I think there are two items yet to be done for this PR:

- [ ] remove the blank line indicated by @rossbar [below](https://github.com/networkx/networkx/pull/5033#discussion_r698050933)
- [ ] Add a check to `multi_source_djikstra_path_length` (which is just about `multi_source_djikstra` in the file).

Comment (divyanx) [reply networkx/algorithms/shortest_paths/weighted.py]: Hey @dschult, I have added the check in `multi_source_djikstra_path_length`,  also sorry for the blank line. Is everything fine now?, can you review it? :)

**Comment by rossbar** [comment]
Ok, the node checks have been added everywhere that calls `_multisource_dijkstra`, but the checks are still in `_multisource_dijkstra` itself. I think the final step is to remove the NodeNotFound checks from `_multisource_dijkstra` (if I'm understanding [dschult's comment](https://github.com/networkx/networkx/pull/5033#discussion_r698068497) correctly)

**Comment by divyanx** [comment]
> Ok, the node checks have been added everywhere that calls `_multisource_dijkstra`, but the checks are still in `_multisource_dijkstra` itself. I think the final step is to remove the NodeNotFound checks from `_multisource_dijkstra` (if I'm understanding [dschult's comment](https://github.com/networkx/networkx/pull/5033#discussion_r698068497) correctly)

Hey @rossbar is that what you meant, please verify.

**Comment by dschult** [comment]
Yes -- thank you @rossbar for your restatement and sorry that my "checklist" is not clear or correct.... :{

There is currently a check in `_multisource_djikstra` two lines that check if the source is in the graph.  Because that is now done in the calling functions it should not be done in the called function.  So those two lines should be removed from `_multicourse_djikstra`.  At least I think that is correct...  Is this more clear?

**Comment by rossbar** [comment]
Just to try to help grease the gears a bit here, I went ahead and made the change we were referring to above in d80785f. This also required slightly modifying the checks in a couple other functions to ensure that the "source not in graph" case was still handled correctly - see 80600fc. I also made some minor updates to the test comments so that they better reflect the current state of things.

@divyanx since I pushed directly to your branch here, please be sure to `git pull`. If you have any trouble (e.g. you get merge conflicts when attempting to pull because you have local changes), please ping here and I'd be happy to help you resolve any problems!

**Comment by rossbar** [comment]
Also @dschult I took the liberty of "dismissing" your initial review since there have been a number of additional changes since the original approval. I just wanted to make sure you had a chance to get eyes on the new changes, I hope you don't mind!

File: networkx/algorithms/shortest_paths/weighted.py
Comment (rossbar) [inline networkx/algorithms/shortest_paths/weighted.py:625]: For the functions where `sources` is a container of nodes, it might be a slight improvement to do something like:

```suggestion
    if set(sources) - set(G):
        raise nx.NodeNotFound(f"Node(s) {set(sources) - set(G)} not found in graph")
```

The reason being, if the user supplies a set of nodes where *more than 1* node isn't in `G`, then this will catch all of the instances rather than iterating through and catching the non-existent nodes one at a time.

This is a very minor consideration and not a blocker here - happy to leave it for a follow-on PR. Just some food for thought!

**Comment by dschult** [comment]
> Also @dschult I took the liberty of "dismissing" your initial review since there have been a number of additional changes since the original approval. I just wanted to make sure you had a chance to get eyes on the new changes, I hope you don't mind!

I don't mind at all -- I have wanted to dismiss my own approval in the past and not found how to do it -- so thanks for this! :}

**Review by dschult — APPROVED** [review]
This looks good to me. :}
Thanks @divyanx and @rossbar

**Review by rossbar — APPROVED** [review]
Okay, I think that covers everything. Thanks @divyanx !

**Pull request merged**
